Quaver.Shared/Screens/Select/SelectScreen.cs

```diff
@@ -536,5 +556,93 @@ public void SelectRandomMap()
                     break;
             }
         }
+
+        /// <summary>
+        ///     Starts the deleting process.
+        /// </summary>
+        private void DeleteSelected()
+        {
+            // Externally loaded map check.
+            if (MapManager.Selected.Value.Game != MapGame.Quaver)
+            {
+                // Display error message.
+                NotificationManager.Show(NotificationLevel.Error, "This map was loaded from another game, and it cannot be deleted.");
+                return;
+            }
+
+            var view = View as SelectScreenView;
+            var type = view.ActiveContainer;
+
+            var selectedMapsetIndex = view.MapsetScrollContainer.SelectedMapsetIndex;
+            var selectedDifficultyIndex = view.DifficultyScrollContainer.SelectedMapIndex;
+            var selectedMapset = AvailableMapsets[selectedMapsetIndex];
+            var selectedDifficulty = selectedMapset.Maps[selectedDifficultyIndex];
+
+            var mapsetPath = Path.Combine(ConfigManager.SongDirectory.Value, selectedMapset.Directory);
+            var difficultyPath = Path.Combine(mapsetPath, selectedDifficulty.Path);
+
+            // Commence deleting and reloading.
+            var confirmDelete = new ConfirmCancelDialog($"Are you sure you want to delete this {( type == SelectContainerStatus.Mapsets ? "mapset" : "difficulty" )}?", (sender, confirm) =>
+            {
+                var mapTitle = selectedMapset.Maps.First().Title;
+                var deleteMapset = type == SelectContainerStatus.Mapsets || type == SelectContainerStatus.Difficulty && selectedMapset.Maps.Count == 1;
+
+                // Dispose of the background for the currently selected map.
+                BackgroundHelper.Background?.Dispose();
+
+                // Dispose of the currently playing track.
+                AudioEngine.Track?.Dispose();
+
+                // Run path deletion in the background.
+                ThreadScheduler.Run(() => DeletePath(deleteMapset ? mapsetPath : difficultyPath));
+
+                // Remove mapset/difficulty from cache and AvailableMapsets list.
+                if (deleteMapset)
+                {
+                    selectedMapset.Maps.ForEach(MapDatabaseCache.RemoveMap);
+                    AvailableMapsets.RemoveAt(selectedMapsetIndex);
+                    MapManager.Mapsets.Remove(selectedMapset);
```

Comment:
This doesn't actually remove the mapset because mapsets in `AvailableMapsets` are different objects than the ones in `MapManager.Mapsets`. This is because searching can select only part of the difficulties in a mapset (so a different `Mapset` object is required).

Reply:
Could probably remove by `selectedMapset.Directory`.